File: test_samples/go/vulnerable.go
package main

import (
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os/exec"
	"strings"
)

// ==================== GO-SEC-001: Command Injection (CRITICAL) ====================

func runCommand(userInput string) {
	cmd := exec.Command("sh", "-c", "ls "+userInput)
	_ = cmd
	// Line: exec.Command with concat
}

// ==================== GO-SEC-002: SQL Injection (CRITICAL) ====================

func queryDB(userName string) {
	query := fmt.Sprintf("SELECT * FROM users WHERE name = '%s'", userName)
	_ = query
	// Line: SQL injection fmt.Sprintf

	query2 := fmt.Sprintf("SELECT * FROM orders WHERE id = '%s'", userName)
	_ = query2
	// Line: SQL concat
}

// ==================== GO-SEC-003: Hardcoded Secrets (HIGH) ====================

const API_KEY = "sk-live-abc123xyz456secret789key12345"
const AWS_KEY = "AKIAIOSFODNN7EXAMPLE"
const GITHUB_TOKEN = "ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"

var dbPassword = "MySecretPass123!"
var jwtSecret = "super_secret_jwt_key"
var authToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4ifQ"

// ==================== GO-SEC-004: Path Traversal (HIGH) ====================

func readFile(userPath string) {
	data, _ := ioutil.ReadFile(userPath + "/data.txt")
	_ = data
	// Line: ioutil.ReadFile path traversal
}

// ==================== GO-SEC-005: Weak Crypto (MEDIUM) ====================

func hashPassword(password string) string {
	h := md5.Sum([]byte(password))
	return fmt.Sprintf("%x", h)
	// Line: MD5 weak crypto
}

func generateToken() string {
	n, _ := rand.Int(rand.Reader, 1000000)
	return fmt.Sprintf("%d", n)
	// Line: rand.Intn not crypto secure
}

func generateRSA() {
	_, _ = rsa.GenerateKey(rand.Reader, 1024)
	// Line: 1024-bit RSA weak
}

// ==================== GO-SEC-006: SSRF (HIGH) ====================

func fetchURL(userURL string) {
	resp, err := http.Get("https://api.example.com/" + userURL)
	if err != nil {
		log.Println(err)
		return
	}
	defer resp.Body.Close()
	_ = resp
	// Line: SSRF http.Get
}

// ==================== GO-SEC-007: Insecure Cookie (MEDIUM) ====================

func setCookie(w http.ResponseWriter) {
	cookie := &http.Cookie{
		Name:   "session",
		Value:  "abc123",
		Secure: false,
	}
	http.SetCookie(w, cookie)
	// Line: insecure cookie Secure=false
}

// ==================== GO-SEC-008: Information Disclosure (LOW) ====================

func logData(password string) {
	log.Println("Password:", password)
	log.Printf("Secret: %s", jwtSecret)
	// Line: log sensitive data
}

// ==================== Clean code (no issues) ====================

func safeQuery(userID string) {
	// Use parameterized queries: db.Query("SELECT * FROM users WHERE id = $1", userID)
}

func safeHash(password string) string {
	h := sha256.Sum256([]byte(password))
	return fmt.Sprintf("%x", h)
}

func safeToken() string {
	b := make([]byte, 32)
	rand.Read(b)
	return fmt.Sprintf("%x", b)
}

func safeRedirect(userURL string) {
	if !strings.HasPrefix(userURL, "https://trusted.com/") {
		return
	}
	_ = userURL
}

File: test_samples/javascript/vulnerable.js
/**
 * Vulnerable JavaScript/TypeScript test file
 * This file intentionally contains security vulnerabilities to test detection.
 * DO NOT use this code in production.
 */

// ==================== JS-SEC-001: Code Injection (CRITICAL) ====================

// eval with user input
const userInput = req.body.code;
eval(userInput);                              // LINE: eval with request body

// new Function with dynamic code
const dynamicFn = new Function('x', 'return x + ' + userInput); // LINE: new Function

// vm.runInContext (Node.js specific)
const vm = require('vm');
const sandbox = { data: userInput };
vm.runInContext('data = data + require("child_process").execSync("ls")', sandbox); // LINE: vm.runInContext

// vm.compileFunction
const compiled = vm.compileFunction('return ' + userInput, [], { parsingContext: sandbox }); // LINE: compileFunction

// ==================== JS-SEC-002: SQL Injection (CRITICAL) ====================

// Template literal in SQL query
const sql = `SELECT * FROM users WHERE name = '${req.query.username}'`; // LINE: SQL template
db.query(sql);

// String concatenation in query
const q = "SELECT * FROM products WHERE id = " + req.params.id; // LINE: SQL concat
connection.execute(q);

// execute with string interpolation
await pool.execute("SELECT * FROM orders WHERE customer = '" + req.body.name + "'"); // LINE: execute concat

// ==================== JS-SEC-003: SSRF (HIGH) ====================

// fetch with template literal URL
const targetUrl = req.query.url;
fetch(`https://api.example.com/data?url=${targetUrl}`); // LINE: fetch SSRF

// axios with dynamic URL
axios.get(`${process.env.API_BASE}/${req.params.endpoint}`); // LINE: axios SSRF

// request library with user URL
const request = require('request');
request({ url: req.body.redirectUrl }, callback); // LINE: request SSRF

// undici with overridable origin
const { request: undiciReq } = require('undici');
undiciReq({ origin: 'http://internal-admin', pathname: '/' + req.query.path }); // LINE: undici SSRF

// ==================== JS-SEC-004: Path Traversal (HIGH) ====================

// readFile with dynamic path
const fs = require('fs');
const file = req.query.filename;
const content = fs.readFileSync(`${file}`, 'utf8'); // LINE: readFile traversal

// createReadStream with user path
fs.createReadStream(req.params.path + '/data.txt'); // LINE: createReadStream traversal

// dynamic require
const moduleName = req.body.module;
const mod = require(`${moduleName}`); // LINE: dynamic require traversal

// dynamic import
import(`${process.cwd()}/modules/${req.query.name}`).then(m => m.run()); // LINE: dynamic import traversal

// ==================== JS-SEC-005: XSS / DOM Injection (HIGH) ====================

// innerHTML with template literal
document.getElementById('output').innerHTML = `<div>${userData}</div>`; // LINE: innerHTML XSS

// innerHTML with concatenation
document.querySelector('.content').innerHTML = '<p>' + req.body.text + '</p>'; // LINE: innerHTML concat XSS

// jQuery .html() with dynamic content
$('#comments').html(req.query.comment); // LINE: jQuery HTML XSS

// React dangerouslySetInnerHTML
return <div dangerouslySetInnerHTML={{ __html: rawUserContent }} />; // LINE: React XSS

// document.write
document.write('<script>' + location.hash.substring(1) + '</script>'); // LINE: document.write XSS

// ==================== JS-SEC-006: Hardcoded Secrets (HIGH) ====================

const API_KEY = "sk-live-abc123xyz456secret789key"; // LINE: hardcoded API key
const AWS_ACCESS_KEY = "AKIAIOSFODNN7EXAMPLE"; // LINE: hardcoded AWS key
const GITHUB_TOKEN = "ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"; // LINE: hardcoded GitHub token
const JWT_SECRET = "super_secret_jwt_key_do_not_use_in_prod"; // LINE: hardcoded JWT secret
const DB_PASSWORD = "MySecretPass123!"; // LINE: hardcoded password
const AUTH_BEARER = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9"; // LINE: hardcoded Bearer token
const PRIVATE_KEY = "-----BEGIN RSA PRIVATE KEY-----\nMIIBOgIBAAJBAL..."; // LINE: hardcoded private key

// ==================== JS-SEC-007: Weak Crypto (MEDIUM) ====================

const crypto = require('crypto');

// MD5 hash
const md5hash = crypto.createHash('md5').update(userInput).digest('hex'); // LINE: MD5

// SHA1 hash
const sha1hash = crypto.createHash('sha1').update(password).digest('hex'); // LINE: SHA1

// Weak DES cipher
const cipher = crypto.createCipher('des', encryptionKey); // LINE: DES

// Math.random for security
const sessionToken = Math.random().toString(36).substring(2); // LINE: Math.random

// JWT with verify disabled
const token = jwt.sign(payload, secret, { algorithm: 'HS256', verify: false }); // LINE: JWT verify disabled

// ==================== JS-SEC-008: Insecure TLS (HIGH) ====================

// HTTPS agent with verify disabled
const agent = new https.Agent({
    rejectUnauthorized: false // LINE: TLS verify disabled
});

// Cookie without secure flag
const cookie = cookieParser.serialize('session', sessionId, {
    httpOnly: true,
    secure: false // LINE: cookie secure flag disabled
});

// ==================== JS-SEC-009: Prototype Pollution (HIGH) ====================

// __proto__ assignment
const userPayload = JSON.parse(req.body.data);
userPayload.__proto__.isAdmin = true; // LINE: proto pollution

// constructor.prototype modification
const config = { ...req.query };
config.constructor.prototype.validate = function() {}; // LINE: constructor prototype

// Object.assign with user data
const merged = Object.assign({}, defaultConfig, req.body); // LINE: Object.assign pollution

// Dynamic key with user input
const key = req.query.propName;
obj[`__${key}__`] = req.body.value; // LINE: dynamic key pollution

// ==================== JS-SEC-010: Information Disclosure (LOW) ====================

// Console log with sensitive data
console.log('Password:', user.password); // LINE: console log password
console.log('Token:', process.env.API_TOKEN); // LINE: console log token
console.log('Full request:', req); // LINE: console log full request

// ==================== Clean code (no issues) ====================

// This should NOT trigger any rules
function sanitize(input) {
    return input.replace(/[<>'"]/g, '');
}

const safeQuery = db.query('SELECT * FROM users WHERE id = ?', [userId]);
const safeUrl = fetch('https://trusted-api.example.com/data');
const safePath = path.resolve('/allowed/dir', req.params.file);
const safeInner = document.getElementById('output').textContent = userData;
const safeKey = process.env.SECURE_API_KEY;
